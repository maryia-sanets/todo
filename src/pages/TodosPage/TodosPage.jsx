import React, { useState, useEffect } from 'react';

import { TodoForm } from '../../components/TodoForm';
import { TodoList } from '../../components/TodoList';

export const TodosPage = () => {
  const [todos, setTodos] = useState([]);

  useEffect(() => {
    async function fetchTodo() {
      const response = await fetch('/todos');
      const json = await response.json();
      setTodos(json);
    }
    fetchTodo();
  }, []);

  const addTodoHandler = async (title) => {
    const response = await fetch('/todos/add', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ title: title })

    });
    const json = await response.json();
    setTodos(json);
  }

  
  const toggleHandler = async (id) => {
    const response = await fetch(`/todos/${id}/update`, {
      method: 'PUT'
    })
    const json = await response.json();
    setTodos(json);
  }

  const deleteHandler = async (id) => {
    const response = await fetch(`/todos/${id}/delete`, {
      method: 'DELETE'
    });
    const json = await response.json();
    setTodos(json);
  }

  return (
    <>
      <TodoForm
        onAdd={addTodoHandler}
      />
      <TodoList
        todos={todos}
        onToggle={toggleHandler}
        onDelete={deleteHandler}
      />
    </>
  )
}