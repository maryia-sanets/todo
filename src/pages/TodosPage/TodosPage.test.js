import React from 'react'

// import API mocking utilities from Mock Service Worker
import { rest } from 'msw'
import { setupServer } from 'msw/node'

// import react-testing methods
import { render, waitFor, screen } from '@testing-library/react'

import '@testing-library/jest-dom/extend-expect'
import ReactDOM from 'react-dom';
import { TodosPage } from './TodosPage';
import { act } from 'react-dom/test-utils'
import userEvent from "@testing-library/user-event";


const server = setupServer(
  rest.get('/todos', (req, res, ctx) => {
    return res(ctx.json([{"id":"anyId","title":"samsepi0l","completed":false}]));
  }),
  rest.post('/todos/add', (req, res, ctx) => {
    return res(ctx.json([{"id":"anyId","title":"my-new-todo","completed":false}]));
  }),
)

let container;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

beforeAll(() => {
  server.listen();
});
afterEach(() => {
  document.body.removeChild(container);
  container = null;
  server.resetHandlers();
});
afterAll(() => {
  server.close();
});


describe('render TodosPage', () => {
  
  it("before loading data from server", () => {
    act(() => {
      ReactDOM.render(<TodosPage />, container);
    });
    const { asFragment } = render(<TodosPage />);
    expect(asFragment(<TodosPage />)).toMatchSnapshot();
  });
  
  it('loading and displaying todos', async () => {
    act(() => {
      ReactDOM.render(<TodosPage url="/todos" />, container);
    });
    await waitFor(() => screen.findByText('TodoList'));
    expect(await screen.findByText(/samsepi0l/i)).toBeInTheDocument();
  });
  
  it('adding todo to page', async () => {
    act(() => {
      ReactDOM.render(<TodosPage url="/todos/add" />, container);
    });
    userEvent.type(screen.getByRole("textbox"), "my-new-todo");
    userEvent.type(screen.getByRole('textbox'), '{enter}');
    expect(await screen.findByText(/my-new-todo/i)).toBeInTheDocument();
  });
  
})






