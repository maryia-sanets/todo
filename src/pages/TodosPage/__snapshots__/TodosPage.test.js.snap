// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render TodosPage after loading data from server and displaying todos 1`] = `
<DocumentFragment>
  <form
    class="form-field"
  >
    <label
      class="form-field-label"
      for="todo"
    >
      New todo
    </label>
    <input
      class="form-field-input"
      placeholder="Enter your todo"
      type="text"
      value=""
    />
  </form>
  <p>
    Your todo list is empty
  </p>
</DocumentFragment>
`;

exports[`render TodosPage before loading data from server 1`] = `
<DocumentFragment>
  <form
    class="form-field"
  >
    <label
      class="form-field-label"
      for="todo"
    >
      New todo
    </label>
    <input
      class="form-field-input"
      placeholder="Enter your todo"
      type="text"
      value=""
    />
  </form>
  <p>
    Your todo list is empty
  </p>
</DocumentFragment>
`;

exports[`render TodosPage before loading data from server 2`] = `
<DocumentFragment>
  <form
    class="form-field"
  >
    <label
      class="form-field-label"
      for="todo"
    >
      New todo
    </label>
    <input
      class="form-field-input"
      placeholder="Enter your todo"
      type="text"
      value=""
    />
  </form>
  <p>
    Your todo list is empty
  </p>
</DocumentFragment>
`;

exports[`render TodosPage loading and displaying todos 1`] = `
<DocumentFragment>
  <form
    class="form-field"
  >
    <label
      class="form-field-label"
      for="todo"
    >
      New todo
    </label>
    <input
      class="form-field-input"
      placeholder="Enter your todo"
      type="text"
      value=""
    />
  </form>
  <p>
    Your todo list is empty
  </p>
</DocumentFragment>
`;
